Kafka/src/main/java/io/deephaven/kafka/publish/JsonKeyOrValueSerializer.java

```diff
@@ -0,0 +1,910 @@
+
+        abstract void processField(FieldContext fieldContext,
+                WritableObjectChunk<ObjectNode, Attributes.Values> jsonChunk, OrderedKeys keys, boolean isRemoval);
+    }
+
+    private class ByteFieldProcessor extends JSONFieldProcessor {
+        private final ColumnSource chunkSource;
+
+        public ByteFieldProcessor(String fieldName, ColumnSource chunkSource) {
+            super(fieldName);
+            this.chunkSource = chunkSource;
+        }
+
+        private class ByteContext implements FieldContext {
+            ChunkSource.GetContext getContext;
+            ByteChunk inputChunk;
+
+            ByteContext(int size) {
+                getContext = chunkSource.makeGetContext(size);
+            }
+
+            @Override
+            public void close() {
+                getContext.close();
+            }
+        }
+
+        @Override
+        FieldContext makeContext(int size) {
+            return new ByteContext(size);
+        }
+
+        @Override
+        void processField(FieldContext fieldContext, WritableObjectChunk<ObjectNode, Attributes.Values> jsonChunk,
+                OrderedKeys keys, boolean previous) {
+            final ByteContext byteContext = (ByteContext) fieldContext;
+            if (previous) {
+                byteContext.inputChunk = chunkSource.getPrevChunk(byteContext.getContext, keys).asByteChunk();
+            } else {
+                byteContext.inputChunk = chunkSource.getChunk(byteContext.getContext, keys).asByteChunk();
+            }
+
+            for (int ii = 0; ii < byteContext.inputChunk.size(); ++ii) {
+                final ObjectNode node = getChildNode(jsonChunk.get(ii));
+                final byte raw = byteContext.inputChunk.get(ii);
+                if (raw == QueryConstants.NULL_BYTE) {
+                    if (outputNulls) {
+                        node.putNull(childNodeFieldName);
+                    }
+                } else {
+                    node.put(childNodeFieldName, raw);
+                }
+            }
+        }
+    }
+
+    private class CharFieldProcessor extends JSONFieldProcessor {
+        private final ColumnSource chunkSource;
+
+        public CharFieldProcessor(String fieldName, ColumnSource chunkSource) {
+            super(fieldName);
+            this.chunkSource = chunkSource;
+        }
+
+        private class CharContext implements FieldContext {
+            ChunkSource.GetContext getContext;
+            CharChunk inputChunk;
+
+            CharContext(int size) {
+                getContext = chunkSource.makeGetContext(size);
+            }
+
+            @Override
+            public void close() {
+                getContext.close();
+            }
+        }
+
+        @Override
+        FieldContext makeContext(int size) {
+            return new CharContext(size);
+        }
+
+        @Override
+        void processField(FieldContext fieldContext, WritableObjectChunk<ObjectNode, Attributes.Values> jsonChunk,
```

Comment:
So we could use replication, but not generics if we want to read the chunk nicely and do the comparisons.  We could also say screw it "JSON is bloated" why not just box the crap and have a single version.

One issue with replication is the inner classes, which would be less beautiful but are obviously doable by just passing in other parameters and having them be somewhat incestuous.

Reply:
Sounds hard and painful.  Let's just leave it as it is.  I did my reviewer duty of asking "maybe this can be avoided without too much work?", the answer was NO.